### ISSUE TITLE
`methods` returns wrong result for triangular dispatch

### ISSUE THREAD

**Issue by yuyichao** [JuliaLang/julia#18348]
See below, `methods` returns two methods even though only the second one (the first in the returned array) can possibly be called. This is the reason we are doing a dynamic dispatch at all in https://github.com/JuliaLang/julia/issues/18331 . (and without this fixed allocation elimination can't fully work either)

``` jl
julia> f{T<:Any}(::Type{T}, x) = 1
f (generic function with 1 method)

julia> f{T<:Any}(::Type{T}, x::T) = 2
f (generic function with 2 methods)

julia> methods(f, Tuple{Type{Any},Any})
#2 methods for generic function "f":
f{T}(::Type{T}, x::T) at REPL[2]:1
f{T}(::Type{T}, x) at REPL[1]:1
```

**Comment by yuyichao** [comment]
Ah....... Just notice this is the first thing @vtjnash mentioned in https://github.com/JuliaLang/julia/issues/18331#issuecomment-244401028 (Didn't realized that since this is also the case on 0.4 and is not the regression........... The regression is still the performance of the dynamic dispatch itself.)

**Comment by JeffBezanson** [comment]
closed by #18457
